frontend/web/components/CollapsibleNestedList.tsx

```diff
@@ -0,0 +1,183 @@
+import React, { useState, forwardRef, useImperativeHandle } from 'react'
+import Icon from 'components/Icon'
+import { EditPermissionsModal } from 'components/EditPermissions'
+import {
+  useGetRoleProjectPermissionsQuery,
+  useGetRoleEnvironmentPermissionsQuery,
+} from 'common/services/useRolePermission'
+import Format from 'common/utils/format'
+
+type MainItem = {
+  name: string
+  id: string
+}
+
+type CollapsibleNestedListProps = {
+  mainItems: MainItem[]
+  role: Role
+  level: string
+}
+
+const CollapsibleNestedList: React.FC<CollapsibleNestedListProps> = forwardRef(
+  ({ level, mainItems, role }, ref) => {
+    const [expandedItems, setExpandedItems] = useState<string[]>([])
+    const [unsavedProjects, setUnsavedProjects] = useState<string[]>([])
+
+    const toggleExpand = (id: string) => {
+      setExpandedItems((prevExpanded) =>
+        prevExpanded.includes(id)
+          ? prevExpanded.filter((item) => item !== id)
+          : [...prevExpanded, id],
+      )
+    }
+
+    const removeUnsavedProject = (projectId) => {
+      setUnsavedProjects((prevUnsavedProjects) =>
+        prevUnsavedProjects.filter((id) => id !== projectId),
+      )
+    }
+
+    useImperativeHandle(
+      ref,
+      () => {
+        return {
+          onClosing() {
+            if (unsavedProjects.length > 0) {
+              return new Promise((resolve) => {
+                openConfirm(
+                  'Are you sure?',
+                  'Closing this will discard your unsaved changes.',
+                  () => resolve(true),
+                  () => resolve(false),
+                  'Ok',
+                  'Cancel',
+                )
+              })
+            } else {
+              return Promise.resolve(true)
+            }
+          },
+          tabChanged() {
+            return unsavedProjects.length > 0
+          },
+        }
+      },
+      [unsavedProjects],
+    )
+
+    const PermissionsSummary = ({ levelId }) => {
```

Comment:
Does this need to be defined within the render function of the parent component?

Reply:
No, Moved.